consul/consul.go

```diff
@@ -137,15 +134,15 @@ func (c *Consul) Register(service *registry.Service) {
 //
 func (c *Consul) Deregister() error {
 	for s, b := range serviceCache {
-		if !b.isRegistered {
+		if c.CacheIsValid(s) {
+			c.CacheProcessDeregister(s)
+		} else {
```

Comment:
If `--heartbeatsBeforeRemove=1`, doesn't this actually give the service two chances? The first time through, `c.CacheIsValid(s)` returns true since `0 <  1` and then the second time through the service is deregistered since `1 !< 1`. Or am I missing something?

Should this be:
```
c.CacheProcessDeregister(s)
if (c.CacheIsValid(s)) {
```

If `--heartbeatsBeforeRemove=1`, doesn't this actually give the service two chances?